spring-cloud-app-broker-logging/src/main/java/org/springframework/cloud/appbroker/logging/streaming/ApplicationLogStreamPublisher.java

```diff
@@ -0,0 +1,147 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+
+import org.springframework.cloud.appbroker.logging.streaming.events.ServiceInstanceLogEvent;
+import org.springframework.cloud.appbroker.logging.streaming.events.ServiceInstanceLoggingEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationListener;
+
+public class ApplicationLogStreamPublisher implements ApplicationListener<ServiceInstanceLoggingEvent> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ApplicationLogStreamPublisher.class);
+
+	private final Map<String, Registration> registry = new HashMap<>();
+
+	private final LogStreamPublisher<org.cloudfoundry.dropsonde.events.Envelope> logStreamPublisher;
+
+	private final ApplicationEventPublisher publisher;
+
+	public ApplicationLogStreamPublisher(
+		LogStreamPublisher<org.cloudfoundry.dropsonde.events.Envelope> logStreamPublisher,
+		ApplicationEventPublisher publisher) {
+		this.logStreamPublisher = logStreamPublisher;
+		this.publisher = publisher;
+	}
+
+	@Override
+	public void onApplicationEvent(ServiceInstanceLoggingEvent event) {
+		final String serviceInstanceId = event.getServiceInstanceId();
+		switch (event.getOperation()) {
+			case START:
+				LOG.debug("Received event to begin listening to logs for {}", serviceInstanceId);
+				this.startPublishing(serviceInstanceId);
+				return;
+			case STOP:
+				LOG.debug("Received event to stop listening to logs for {}", serviceInstanceId);
+				this.stopPublishing(serviceInstanceId);
+				return;
+
+			default:
+				throw new IllegalArgumentException("Unknown operation: " + event.getOperation());
+		}
+	}
+
+	private void startPublishing(String serviceInstanceId) {
+		synchronized (registry) {
+			final Registration registration = registry.get(serviceInstanceId);
+			if (registration != null) {
+				LOG.debug("Incrementing registration subscription count for {}", serviceInstanceId);
+				registration.increment();
+
+				return;
+			}
+
+			Flux<org.cloudfoundry.dropsonde.events.Envelope> logStream = this.logStreamPublisher
+				.getLogStream(serviceInstanceId);
+
+			final Disposable subscription = logStream
+				.doOnNext(
+					envelope -> publisher
+						.publishEvent(new ServiceInstanceLogEvent(this, serviceInstanceId, envelope)))
+				.subscribe();
+
+			LOG.debug("Creating new registration for {}", serviceInstanceId);
+			registry.put(serviceInstanceId, new Registration(subscription));
+		}
+	}
+
+	private void stopPublishing(String serviceInstanceId) {
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Received event to stop publishing logs for {}", serviceInstanceId);
+		}
+
+		synchronized (registry) {
+			final Registration registration = registry.get(serviceInstanceId);
+			if (registration == null) {
+				if (LOG.isWarnEnabled()) {
```

Comment:
```suggestion
```

Reply:
I mean... in this case this just adding a redundant check. It's _more_ work.

Reply:
Actually, the compiler will swallow that and move on. But I agree it's extra noise to the readers' eyes.
Happy to recondiser the removal of the PMD rules but as of this PR, I think we should leave them for consistency.

Reply:
Ah you're right of course. And yes let's not get into changing the PMD rules. I forgot we had them.